### DISCUSSION TITLE
Prevent `GameUpdate` ticks triggered by non-terraria connections (like bots and scripts).

### DISCUSSION THREAD

**Discussion by pcider** [Pryaxis/TShock#2600] [Ideas]
Almost always, if you have a terraria server exposed to the internet, there will be bots which trigger the server's `GameUpdate` while the bot's connection is still alive. This happens because `Netplay.HasClients` is set to `true` even if the client hasn't sent a single game packet. More annoying, they also trigger backups and weird things happening like bee queen spawning in a bee world (probably due to a falling star hitting the hive when the game ticks) even when there are no players online (2nd Picture).  

I'm wondering if it is better to suppress game updates and backups for when these bots are connecting to the server. Suppressing game update ticks would require OTAPI to be able to suppress the `Main.Update` method. Is this feature safe and sound to implement?

![image](https://user-images.githubusercontent.com/28196972/158946267-86306ec5-8e37-408a-9735-1dd3788f53b5.png)
![image](https://user-images.githubusercontent.com/28196972/158947456-b8fa33a2-d6a3-4970-bfa3-8ec6a6c813b7.png)
(xPanini's Terraria-Chat-Relay)

**Comment by hakusaro** [comment]
Imho this would be better to address in the base game, but it’s probably a good use case for an OTAPI patch.

**Reply by bartico6** [reply]
It would be better, but it's not going to happen xD

~~Simple fix:~~ read below

```csharp
        internal void AttachTickHook()
        {
            Main.OnTickForThirdPartySoftwareOnly += () =>
            {
                Netplay.HasClients = Main.player.Any(p => p != null && p.active && Netplay.Clients[p.whoAmI].State == 10);
            };
        }
```

will make world ticks only happen if there's at least one player fully loaded into the server.

(yes it's that simple, yes I do it myself in my server fork, yes it works)

((but you should still test it because of the multitude of changes to netcode I've done in my server, this is a compat bridge I use to make it recognize what my netcode is doing))

(((this won't work out of the box by itself actually since packet handling happens on main thread which needs to be woken up from permasleep by setting hasclients, there's plenty of additional work to be done but this is a starter for it all, sorry for the initially misleading shitpost)))

**Reply by hakusaro** [reply]
@bartico6 apology accepted

**Reply by pcider** [reply]
One tick would still be fired when using the `OnTickForThirdPartySoftwareOnly` hook since the check for `Netplay.HasClients` occurs before that. I was thinking of using the OTAPI's `OTAPI.Hooks.Game.StatusTextUpdate` hook since it happens just before the call to `Main.Update` in the `DedServ` method.

```c#
Hooks.Game.StatusTextUpdate += () =>
{
    Netplay.HasClients = Main.player.Any(p => p != null && p.active && Netplay.Clients[p.whoAmI].State == 10);
    
    // Handle Netplay stuff here
    if (!Netplay.HasClients)
    {
        Netplay.UpdateInMainThread();
        for (int i = 0; i < 256; i++)
        {
            if (Netplay.Clients[i].Socket != null)
            {
                Netplay.Clients[i].Socket.SendQueuedPackets();
            }
        }
    }
}
```

What do you think?

**Reply by drunderscore** [reply]
@gohjoseph Well as particles said in the parent comment
> but it’s probably a good use case for an OTAPI patch.

Which I agree with, the patch would just be the most sensible way, and would prevent one tick from occurring.
